internal/services/notification/runeventssender.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2023 Sorint.lab
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notification
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"sort"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+	"github.com/sorintlab/errors"
+
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/internal/util"
+	"agola.io/agola/services/runservice/types"
+)
+
+type runEventsSender struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	started bool
+
+	log zerolog.Logger
+
+	t *testing.T
+
+	mu sync.Mutex
+
+	listenAddress string
+	exposedURL    string
+
+	runEvents *runEvents
+}
+
+type runEvents struct {
+	runEvents []*types.RunEvent
+	mu        sync.Mutex
+}
+
+func (r *runEvents) addRunEvent(runEvent *types.RunEvent) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.runEvents = append(r.runEvents, runEvent)
+}
+
+func setupRunEventsSender(pctx context.Context, t *testing.T) *runEventsSender {
+	log := testutil.NewLogger(t)
+	ctx, cancel := context.WithCancel(pctx)
+
+	r := &runEvents{runEvents: make([]*types.RunEvent, 0)}
+	es := &runEventsSender{ctx: ctx, t: t, log: log, cancel: cancel, runEvents: r}
+
+	port, err := testutil.GetFreePort("localhost", true, false)
+	if err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	es.listenAddress = net.JoinHostPort("localhost", port)
+	es.exposedURL = fmt.Sprintf("http://%s", net.JoinHostPort("localhost", port))
+
+	if err := es.start(); err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	go func() {
+		<-ctx.Done()
+
+		es.stop()
+	}()
+
+	return es
+}
+
+func (es *runEventsSender) start() error {
+	es.mu.Lock()
+	defer es.mu.Unlock()
+
+	if es.started {
+		return fmt.Errorf("runEventsSender already started")
+	}
+
+	router := mux.NewRouter()
+
+	router.Handle("/api/v1alpha/runs/events", newRunEventsHandler(es.log, es.runEvents)).Methods("GET")
+
+	httpServer := http.Server{
+		Addr:    es.listenAddress,
+		Handler: router,
+	}
+
+	lerrCh := make(chan error)
+	go func() {
```

Comment:
This will create flake errors in test since it starts listening asyncronously. I fixed the same issue for webhooksreceiver in PR #426 .